### ISSUE TITLE
Parameter Array modifier lost in mock

### ISSUE THREAD

**Issue by afscrome** [moq/moq4#235]
The only way to detect through reflection if a method parameter is a parameter array is by checking if the parameter has the `ParamArrayAttribute` attribute defined on it.  However when you mock a method with a params array, the mock implementation looses this attribute.

This behavior can be shown by the below test - I'd expect the last assertion to pass, but it currently fails.

``` csharp
public abstract class Base
{
    public abstract string ParamsArray(params int[] args);
}

[Test]
public void ParamsArrayLostOnMock()
{
    var baseMethodParam = typeof(Base).GetMethods()[0].GetParameters()[0];

    var mock = new Mock<Base>();
    mock.Setup(x => x.ParamsArray(1, 2, 3)).Returns("foo");

    var mockMethodParam = mock.Object.GetType().GetMethods()[0].GetParameters()[0];

    Assert.IsTrue(baseMethodParam.IsDefined(typeof(ParamArrayAttribute), false), "ParamArrayAttribute is not defined on base");
    Assert.IsTrue(mockMethodParam.IsDefined(typeof(ParamArrayAttribute), false), "ParamArrayAttribute is not defined on mock");
}
```

**Comment by kzu** [comment]
This is an issue in Castle.DynamicProxy most likely. Can you see if it can be reproduced with any of the other open source mocking libraries? (I.e. Rhino, NSubstitute, FakeItEasy, etc.?)

**Comment by afscrome** [comment]
I was able to reproduce with NSubstitue, as well as using `ProxyGenerator` directly - see  castleproject/Core#121 .

**Comment by TheFastCat** [comment]
I think I'm also experiencing this... (though they used to work why is that) as Setup()s and Verify(...) calls such as the following

**Comment by kzu** [comment]
Marking as an enhacement that has to happen in Castle.

**Comment by stakx** [comment]
A fix for this is going to be in Castle Core v4.1 (see https://github.com/castleproject/Core/issues/121#issuecomment-307610919). As soon as we upgrade Castle and republish Moq, this can be closed.
